### DISCUSSION TITLE
Crop an object to a square viewport

### DISCUSSION THREAD

**Discussion by mamhlav** [ImageMagick/ImageMagick#2354] [Help]
Hi,
please help me!

I need to detect an object in a photo and cut it out in a square format.

When I use it
"convert" * .jpg "-fuzz 20% -trim -set filename: f% t trimmed /% [filename: f] .jpg"
the object is detected correctly and accurately cropped.

After detecting the object, I need to add 10 px at the top and bottom and make a square cutout.

I enclose the original and the output I need.

Thank you very much for your help!
![original](https://user-images.githubusercontent.com/69124356/89133251-b7ef4a00-d51a-11ea-82ed-87017389d18b.jpg)
![I_need](https://user-images.githubusercontent.com/69124356/89133400-d275f300-d51b-11ea-8759-8a182d65ea52.jpg)

**Comment by snibgo** [comment]
What version of IM, on what platform?

I think what you want is:

1. Find what the trim parameters would be.

2. Add 10 pixels to all four sides.

3. Add pixels to left and right, or top and bottom, as needed to make the image square.

Is that what you want?

I would do this with "-format %@" for step (1), and do the other steps in a shell script. I expect it can be done in a single (messy) IM command.

**Reply by mamhlav** [reply]
Hi, thank you very much for your reply!
I've been working on it all day, but it doesn't work.
Can you clarify this for me, please?

Thank you. Very much.

**Comment by fmw42** [comment]
Here is one way in Unix syntax:

```
Get the color of the top left pixel
Get the crop values from %@
Extract the width and height
Compute the width+20 and height+20 and the maximum of the two
Crop the image and extend it with the background color of the max size previously computed in both width and height

```
```
color=`convert dress.jpg -auto-orient -format "%[pixel:u.p{0,0}]" info:`
cropvals=`convert dress.jpg -auto-orient -fuzz 15% -format "%@" info:`
ww=`echo $cropvals | tr "x" "+" | cut -d+ -f1`
hh=`echo $cropvals | tr "x" "+" | cut -d+ -f2`
wh=`convert xc: -format "%[fx:max($ww+20,$hh+20)]" info:`
convert dress.jpg -auto-orient -crop $cropvals +repage -background $color -gravity center -extent ${wh}x${wh} dress_proc.jpg

```


![dress_proc](https://user-images.githubusercontent.com/13458218/89209823-e2b9bc80-d573-11ea-9f5b-28647a8e3687.jpg)

**Comment by fmw42** [comment]
Here is a better method that keeps the image texture if possible.

```
color=`convert dress.jpg -auto-orient -format "%[pixel:u.p{0,0}]" info:`
WxH=`convert dress.jpg -auto-orient -format "%wx%h" info:`
W=`echo $WxH | cut -dx -f1`
H=`echo $WxH | cut -dx -f2`
cropvals=`convert dress.jpg -auto-orient -fuzz 15% -format "%@" info: | tr "x" "+"`
ww=`echo $cropvals | cut -d+ -f1`
hh=`echo $cropvals | cut -d+ -f2`
xo=`echo $cropvals | cut -d+ -f3`
yo=`echo $cropvals | cut -d+ -f4`
www=$((ww+20))
hhh=$((hh+20))
wh=`convert xc: -format "%[fx:max($www,$hhh)]" info:`
xxo=`convert xc: -format "%[fx:min($xo-10,round(($W-$wh)/2))]" info:`
yyo=`convert xc: -format "%[fx:min($yo-10,round(($H-$wh)/2))]" info:`
convert dress.jpg -auto-orient -crop ${wh}x${wh}+${xxo}+${yy0} +repage -background $color -gravity center -extent ${wh}x${wh} dress_proc2.jpg

```

![dress_proc2](https://user-images.githubusercontent.com/13458218/89212307-15fe4a80-d578-11ea-84fb-67f13116dea6.jpg)

**Reply by mamhlav** [reply]
Hi,
Thank you very much, this is exactly what I need!
Can this be done in Windows as well?
What Linux distribution do you recommend for Imagemagick? I have little experience with CentOS and Ubuntu.

Thanks again for the help !!!
